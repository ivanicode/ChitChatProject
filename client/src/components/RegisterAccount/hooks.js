import {useState} from 'react';
import dayjs from 'dayjs';

export function useAllHooks() {
    const {
        formData, 
        onChangeHandler
    } = useManageFormData();

    const {
        isPasswordValid, 
        inputRepeatedPassword,
        submitForm
    } = useManagePasswordMatch(formData);

    const {
        errors,
        onBirthDateChangeHandler,
    } = useManageErrors(onChangeHandler)

    return {
        isPasswordValid, 
        inputRepeatedPassword,
        submitForm,
        errors,
        onBirthDateChangeHandler,
        formData,
        onChangeHandler,
    }
}

export function useManagePasswordMatch(formData) {
    const [isPasswordValid, setIsPasswordValid] = useState(true);

    function checkIfPasswordMatch(){
        
        return formData.originalPassword === formData.repeatedPassword;
    }
    function showError(){
        setIsPasswordValid(false) 
    }
    function inputRepeatedPassword() {
     
        if (checkIfPasswordMatch()) {
            setIsPasswordValid(true)
            
        } else {
          
          showError();
        }
      }
      function submitForm(event) {
      
        if(!checkIfPasswordMatch()) {
          showError();
          event.preventDefault();
        }
      }
      return {
          isPasswordValid, 
          inputRepeatedPassword,
        submitForm
      }
}

export const initialState = {
    originalPassword: '',
    repeatedPassword: '',
    mail: '',
    date: '',
    lastName: '',
    firstName: ''
}
export function useManageFormData(){

    const [formData, setFormData] = useState(initialState);

    function onChangeHandler(event){
        const id = event.target.id;
        const value = event.target.value;
        const newState = {...formData};
        newState[id] = value;
        setFormData(newState);
    }

    return {formData, onChangeHandler}
}


export function useManageErrors(onChangeHandler){

    const [errors, setErrors] = useState({});

    function onBirthDateChangeHandler(event) {
        onChangeHandler(event);
        const date = dayjs(event.target.value);
        if(!date.isValid()){
            setErrors({
                ...errors, 
                [event.target.id]: 'Data jest niepoprawna',
            });
            return;
        }
        if(date.year() > 2004){
            setErrors({
                ...errors, 
                [event.target.id]: 'Jesteś za młody/a!',
            });
            return;
        }
        const newErrors = {...errors};
        delete newErrors[event.target.id];
        setErrors(newErrors);
    }
    return {errors, onBirthDateChangeHandler}

}