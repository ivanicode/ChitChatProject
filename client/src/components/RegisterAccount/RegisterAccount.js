import React from 'react';
import ErrorText from '../common/ErrorText/ErrorText';
import { useAllHooks } from './registerAccountHooks';
import {useFetch} from '../../common/hooks/useFetchHook'


export function RegisterAccount(){

    const {
        submitForm,
        errors,
        onBirthDateChangeHandler,
        formData,
        onChangeHandler,
        onNameChangeHandler,
        onPasswordBlurHandler,
        onEmailBlurHandler,
        formIsValid,
        onRepeatedPasswordChangeHandler
    } = useAllHooks()

    return (
        <form onSubmit={submitForm} noValidate>
        <div className="registerAccount">
            <label className="firstName" className="registerAccountLabels">
                Imię:
                <input type="text" id="firstName" value={formData.firstName} onChange={onNameChangeHandler} ></input>
                {errors.firstName && <ErrorText error={errors.firstName} />}
            </label>
            <label className="lastName" className="registerAccountLabels">
                Nazwisko:
                <input type="text" id="lastName" value={formData.lastName} onChange={onNameChangeHandler} ></input>
                {errors.lastName && <ErrorText error={errors.lastName} />}
            </label>
            <label className="birthDate" className="registerAccountLabels">
                Data urodzenia:
                <input type="date" max="2004-12-31" min="1900-01-01"
                id="date" value={formData.date} onChange={onBirthDateChangeHandler} onKeyPress={onBirthDateChangeHandler} ></input>
                {errors.date && <ErrorText error={errors.date} />}
            </label>
            <label className="email" className="registerAccountLabels">
                e-mail:
                <input type="email" id="mail" value={formData.mail} onBlur={onEmailBlurHandler} onChange={onChangeHandler}></input>
                {errors.mail && <ErrorText error={errors.mail} />}
            </label>
            <label className="password" className="registerAccountLabels">
                Hasło:
                <input type="password" id="originalPassword" value={formData.originalPassword} onBlur={onPasswordBlurHandler} onChange={onChangeHandler} ></input>
                {errors.originalPassword && <ErrorText error={errors.originalPassword} />}
            </label>
            <label className="repeatPassword" className="registerAccountLabels">
                Powtórz hasło:
                <input type="password" id="repeatedPassword" value={formData.repeatedPassword} onChange={onRepeatedPasswordChangeHandler} ></input>
                {errors.repeatedPassword && <ErrorText error={errors.repeatedPassword} />}
            </label>
            <div className="createAccountButtonDiv">
                <button className="createAccountButton" type="button" disabled={!formIsValid} onClick={submitForm}>
                Stwórz konto
                </button>
            </div>
        </div>
        </form>
    )
}

RegisterAccount.displayName = 'RegisterAccount';
RegisterAccount.propTypes = {};

export default RegisterAccount;




