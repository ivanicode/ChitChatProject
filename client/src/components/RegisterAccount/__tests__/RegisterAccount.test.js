import React from 'react';
import {shallow} from 'enzyme';
import {RegisterAccount} from '../RegisterAccount';
import * as hooks from '../registerAccountHooks';



describe('RegisterAccount component', () => {
    it('should match snapshot when no errors', () => {
        hooks.useAllHooks = jest.fn().mockImplementation(
            () => ({
                errors: {},
                formData: {},
                isPasswordValid: false,
                inputRepeatedPassword: jest.fn(),
                submitForm: jest.fn(),
                onBirthDateChangeHandler: jest.fn(),
                onChangeHandler: jest.fn(),
                onNameChangeHandler: jest.fn(),
                onPasswordBlurHandler: jest.fn(),
                onEmailBlurHandler: jest.fn(),
                onRepeatedPasswordChangeHandler: jest.fn()
            })
        )
        const wrapper = shallow(<RegisterAccount />)
        expect(wrapper).toMatchSnapshot()
    })
    it('should match snapshot when errors exist', () => {
        hooks.useAllHooks = jest.fn().mockImplementation(
            () => ({
                errors: {firstName: 1, lastName: 1, date: 1, mail: 1, originalPassword: 1, repeatedPassword: 1},
                formData: {},
                isPasswordValid: false,
                inputRepeatedPassword: jest.fn(),
                submitForm: jest.fn(),
                onBirthDateChangeHandler: jest.fn(),
                onChangeHandler: jest.fn(),
                onNameChangeHandler: jest.fn(),
                onPasswordBlurHandler: jest.fn(),
                onEmailBlurHandler: jest.fn(),
                onRepeatedPasswordChangeHandler: jest.fn()
            })
        )
        const wrapper = shallow(<RegisterAccount />)
        expect(wrapper).toMatchSnapshot()
    })
})
