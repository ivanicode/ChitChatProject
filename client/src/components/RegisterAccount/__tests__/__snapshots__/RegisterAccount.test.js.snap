// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterAccount component should match snapshot when errors exist 1`] = `
<form
  noValidate={true}
  onSubmit={[MockFunction]}
>
  <div
    className="registerAccount"
  >
    <label
      className="registerAccountLabels"
    >
      Imię:
      <input
        id="firstName"
        onChange={[MockFunction]}
        type="text"
      />
      <ErrorText
        error={1}
      />
    </label>
    <label
      className="registerAccountLabels"
    >
      Nazwisko:
      <input
        id="lastName"
        onChange={[MockFunction]}
        type="text"
      />
      <ErrorText
        error={1}
      />
    </label>
    <label
      className="registerAccountLabels"
    >
      Data urodzenia:
      <input
        id="date"
        max="2004-12-31"
        min="1900-01-01"
        onChange={[MockFunction]}
        onKeyPress={[MockFunction]}
        type="date"
      />
      <ErrorText
        error={1}
      />
    </label>
    <label
      className="registerAccountLabels"
    >
      e-mail:
      <input
        id="mail"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        type="email"
      />
      <ErrorText
        error={1}
      />
    </label>
    <label
      className="registerAccountLabels"
    >
      Hasło:
      <input
        id="originalPassword"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        type="password"
      />
      <ErrorText
        error={1}
      />
    </label>
    <label
      className="registerAccountLabels"
    >
      Powtórz hasło:
      <input
        id="repeatedPassword"
        onChange={[MockFunction]}
        type="password"
      />
      <ErrorText
        error={1}
      />
    </label>
    <div
      className="createAccountButtonDiv"
    >
      <button
        className="createAccountButton"
        disabled={true}
        onClick={[MockFunction]}
        type="button"
      >
        Stwórz konto
      </button>
    </div>
  </div>
</form>
`;

exports[`RegisterAccount component should match snapshot when no errors 1`] = `
<form
  noValidate={true}
  onSubmit={[MockFunction]}
>
  <div
    className="registerAccount"
  >
    <label
      className="registerAccountLabels"
    >
      Imię:
      <input
        id="firstName"
        onChange={[MockFunction]}
        type="text"
      />
    </label>
    <label
      className="registerAccountLabels"
    >
      Nazwisko:
      <input
        id="lastName"
        onChange={[MockFunction]}
        type="text"
      />
    </label>
    <label
      className="registerAccountLabels"
    >
      Data urodzenia:
      <input
        id="date"
        max="2004-12-31"
        min="1900-01-01"
        onChange={[MockFunction]}
        onKeyPress={[MockFunction]}
        type="date"
      />
    </label>
    <label
      className="registerAccountLabels"
    >
      e-mail:
      <input
        id="mail"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        type="email"
      />
    </label>
    <label
      className="registerAccountLabels"
    >
      Hasło:
      <input
        id="originalPassword"
        onBlur={[MockFunction]}
        onChange={[MockFunction]}
        type="password"
      />
    </label>
    <label
      className="registerAccountLabels"
    >
      Powtórz hasło:
      <input
        id="repeatedPassword"
        onChange={[MockFunction]}
        type="password"
      />
    </label>
    <div
      className="createAccountButtonDiv"
    >
      <button
        className="createAccountButton"
        disabled={true}
        onClick={[MockFunction]}
        type="button"
      >
        Stwórz konto
      </button>
    </div>
  </div>
</form>
`;
